[WGSL] Validation of array element type is incorrect

tadeuzagallo opened the pull request:
#### 1b0f8896c4d1d6503e3f560b9bad212aab613763
<pre>
[WGSL] Validation of array element type is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=269386">https://bugs.webkit.org/show_bug.cgi?id=269386</a>
<a href="https://rdar.apple.com/122960489">rdar://122960489</a>

Reviewed by Mike Wyrzykowski.

In 274617@main I added a check that array elements must be constructible, but I missed
that this is only the case when constructing an array. For an array type (e.g. x: array&lt;T&gt;)
the element type must only have creation fixed footprint.

* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):

Canonical link: <a href="https://commits.webkit.org/274718@main">https://commits.webkit.org/274718@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a006a03efce64141dbc8bcb2a54785196210ff80

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39694 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18673 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42050 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42229 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35594 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21573 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16002 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33125 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40268 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15751 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34345 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13653 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13648 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43506 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36041 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35631 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39414 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14506 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11946 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37688 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16112 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8934 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16161 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a006a03efce64141dbc8bcb2a54785196210ff80. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24436

mdubet commented:
Should this have a test case (or improve some already existing test result) ?

mwyrzykowski commented:
> Should this have a test case (or improve some already existing test result) ?

An existing CTS test case covers this, e.g., https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,compute_pipeline,*

That test is marked all passing expectations, so we don't need to update the expectations. We should enable that test in EWS (it's marked Skip currently), that will likely happen in two weeks

tadeuzagallo commented:
These are all covered by the CTS tests.

tadeuzagallo committed:
[WGSL] Validation of array element type is incorrect
https://bugs.webkit.org/show_bug.cgi?id=269386
rdar://122960489

Reviewed by Mike Wyrzykowski.

In 274617@main I added a check that array elements must be constructible, but I missed
that this is only the case when constructing an array. For an array type (e.g. x: array<T>)
the element type must only have creation fixed footprint.

* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):

Canonical link: https://commits.webkit.org/274718@main
